File: problem-set-2/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <math.h>


int get_index(string text);
float get_number_of_letters(string text);
float get_number_of_words(string text);
float get_number_of_sentences(string text);


int main(void)
{
    string text = get_string("Text: ");
    int grade = get_index(text);
    if (grade == 10)
    {
        printf("Grade: X");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1")
    }
    else
    {
        printf("Grade: %i\n", grade);
    }
}


int get_index(string text)
{
    // index = 0.0588 * L - 0.296 * S - 15.8
    float number_of_letters = get_number_of_letters(text);
    float number_of_words = get_number_of_words(text);
    float number_of_sentences = get_number_of_sentences(text);

    // average number of letters per 100 words
    float l = 100.0f *  number_of_letters / number_of_words;

    // average number of sentences per 100 words
    float s = 100.0f * number_of_sentences / number_of_words ;

    float index =0.0588 * l - 0.296 * s - 15.8;
    return round(index);
}

float get_number_of_letters(string text)
{
    int counter = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] >= 'A' && text[i] <= 'z')
       {
            counter++;
        }
    }
    return counter;
}


float get_number_of_words(string text)
{
    int counter = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ' || text[i] == '\n')
        {
            counter++;
        }
    }
    return counter+1;
}


float get_number_of_sentences(string text)
{
    int counter = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if ((int) text[i] == 33 || (int) text[i] == 46 || (int) text[i] == 63)
        {
            counter++;
        }
    }
    return counter;
}

File: problem-set-6/sentimental-readability/readability.py
def main() -> None:
    text = input("Text: ")
    grade = get_index(text)
    if grade >= 16:
        print("Grade 16+")
    elif grade < 1:
        print("Before Grade 1")
    else:
        print(f"Grade: {grade}")


def get_index(text: str) -> int:
    letters = get_letters_number(text)
    words = get_words_number(text)
    sentences = get_sentences_number(text)

    average_letters = 100.0 * letters / words
    average_sentences = 100.0 * sentences / words
    index = 0.0588 * average_letters - 0.296 * average_sentences - 15.8
    return round(index)

    
def get_letters_number(text: str) -> float:
    counter = 0
    for char in text:
        if char.isalpha():
            counter += 1
    return counter


def get_words_number(text: str) -> float:
    counter = 0
    for char in text:
        if char.isspace():
            counter += 1
    return counter + 1


def get_sentences_number(text: str) -> float:
    counter = 0
    for char in text:
        if char == '.' or char == '!' or char == '?':
            counter += 1
    return counter


if __name__ == "__main__":
    main()